### ISSUE TITLE
How do I use `setuptools_scm` with Docker?

### ISSUE THREAD

**Issue by mrmachine** [pypa/setuptools-scm#77]
I want to copy my working copy and `pip install -e .`, without the `.git` folder in the resulting image. But without the `.git` folder, `setuptools_scm` complains because it can't work out the version number to use.

**Comment by jaraco** [comment]
I don't know the answer, but I would try one or two things: 
- run setup.py egg_info on the project before copying without the .git metadata.
- create an sdist `setup.py sdist` and use the sdist (or expanded version thereof) instead of the raw files from the repo. 

My guess/hope is that one of these operations will capture the version number from git.

**Comment by mrmachine** [comment]
I'll give that `egg_info` suggestion a shot. In the meantime, I've worked around it with:

``` python
# Allow installation without git repository, e.g. inside Docker.
if os.path.exists('.git'):
    kwargs = dict(
        use_scm_version={'version_scheme': 'post-release'},
        setup_requires=['setuptools_scm'],
    )
else:
    kwargs = dict(version='0+d'+datetime.date.today().strftime('%Y%m%d'))

setuptools.setup(
    ...
    **kwargs
)
```

This just gives me a date stamped local version from 0, which isn't really accurate but at least it builds and gives me a ballpark idea about when the code came from.

I thought about using an sdist, but that's a bit of a PITA because I'd have to copy the sdist into the docker image in order to install it (leaving me with a redundant file in the image) or make it available over HTTP and pass the URL in as a build arg (more error-prone manual steps, or steps to automate) and re-build a new sdist for every update before I can build the docker image.

It's also a bit of a departure in terms of configuration of the app, as it normally is installed into the venv with `pip install -e .` and it writes a bunch of files to that location (logs, uploaded media, etc.) that are also referenced by `nginx` configs, etc. I'd not like to have to go hunting in the `site-packages` folder for those things.

I had thought it would be nice if `setuptools_scm` did something similar to what is described at: http://blogs.nopcode.org/brainstorm/2013/05/20/pragmatic-python-versioning-via-setuptools-and-git-tags/

Basically, write the dynamic version number to a file whenever `setuptools_scm` generates it, and fallback to that file when the git repository is not found. The local version file would not be committed to source control, but would be included in a built sdist and locally.

**Comment by RonnyPfannschmidt** [comment]
We already support version files, just not as source of truth

From my POV a good way to go would be to support local egg info as fallback source of truth

As for docker usage, I would suggest to use mounts for the file sources and temporary folders, so you won't clutter the image

**Comment by mrmachine** [comment]
Could you elaborate on using mounts for file sources with docker?

The thing that gives me pause about using local egg info as a fallback source of truth is that as soon as you change one file and don't rebuild the local egg info, it's out of date and stale anyway. If it's not absolutely reliable, it's not really any better than just a date stamp or `0+dYYYYMMDD`.

I'd rather `setuptools_scm` maybe the local egg info as a baseline if no `.git` repository is present, and add a `.dYYYYMMDD` local version label segment, just like it does when the working copy is dirty, and ideally, `setuptools_scm` would write this local egg info every time `python setup.py` is run, to ensure it is kept as up to date as possible.

So for example, say I have a commit `1a2b3c4` and its parent is tagged 0.2, `python setup.py --version` might return `0.2.post0.ng1a2b3c4`, and should write the package info. Then that is copied into docker, or the `.git` directory is deleted. Next time `python setup.py ...` is run, `setuptools_scm` would see there is no `.git` repository, use the package info, and append a date stamp, e.g. `0.2.post0.ng1a2b3c4.d20160118`.

I suppose `setuptools_scm` could store a hash of all the package file contents, and omit the date stamp segment if the file contents are an exact match, otherwise it is dirty and the date stamp segment is consistent with the behaviour when `.git` is present.

**Comment by RonnyPfannschmidt** [comment]
The idea of using egg-info as fallback is all about only using it when something liked git is not available

I consider it the task of the packager to script things so the metadata is not stale

The only metadata file we can reliably keep up2date is the cache-file

**Comment by jbarlow83** [comment]
I ran into this issue not by Docker but simply by trying to get a single version number whether my package is install via pip or as a git clone – not a complex case at all. I think it's pretty clear that this package needs to be able to produce the correct version number if the SCM is not present.

**Comment by jbarlow83** [comment]
I think it would be helpful if write_to generated a version.py that was capable of automatically figuring out the version dynamically. For example, try importing setuptools_scm and using SCM, and then fallback on the stored version with some kind of sanity check on the stored version. Then a program that needs to report its own version (e.g. command line tool --version) can import one file and trust its version is correct.

**Comment by RonnyPfannschmidt** [comment]
@jbarlow83 please use pkg-ressources

**Comment by mrmachine** [comment]
How? To get saved version from egg info when setuptools_scm fails? But it fails inside execution of setup(), so we have to add code to predict failure and not use setuptools_scm in that case? Can't setuptools_scm just fallback automatically? Or am I not understanding how we would use pkg resources?

**Comment by jbarlow83** [comment]
I can do this in my command line script (which is part of the package) and it seems to work.

``` python
import pkg_resources
VERSION = pkg_resources.get_distribution('mypackage').version
```

That's pretty much what I want, in the sense that if I install my package into a test virtual environment it picks up the right version. However, in my development environment, it won't pickup the right version. I would have to do something like:

``` python
try:
    # Development environment
    import setuptools_scm
    try:
        # [... get version from setuptools scm ... ]
    except NoSCMAvailableBut_setuptools_scm_HappensToBeInstalledInThisVirtualenvError:
        # @mrmachine 's problem
except ImportError:
   # Production environment
   import pkg_resources
   # [... get version from package resources ... ]

```

Sure, I can do all that, but in my view if I'm bringing a new dependent package into my project for the relativity simple task of getting version numbers, I'd much prefer to see it take care of those cases for me.

**Comment by RonnyPfannschmidt** [comment]
@mrmachine that is the case I yet need to implement

@jbarlow83 its a reasonable downside that dev envs don't get the correct version metadata eagerly
Correct implementation within the constraints of python/setuptools seems to be a relatively large cost at little gain

But I wouldn't mind a good PR showing that I was wrong

**Comment by mrmachine** [comment]
@RonnyPfannschmidt I agree it's reasonable that we don't get the correct version eagerly... However, I think `setuptools_scm` should still return _some_ version number when `setup.py` is run without access to the version control system, instead of crashing and burning and requiring users to dirty up their setup.py with `try/except` blocks and code that tries to predict `setuptools_scm` failure conditions, since we can't actually import `setuptools_scm` and call it in a `try/except` block to detect actual failure.

Without access to VCS, there's no way to know if the stored version in egg info is stale or not anyway, so I think the public version identifier should either be `0` or use public version identifier from the stored version in egg info, but add or replace the local version label, e.g. `+dYYYYMMDD` or `+nh{hash-of-package-contents}.dYYYYMMDD`.

I think building or installing editable from source without existing package info and access to VCS can be considered similar to building or installing from a dirty VCS working copy, for which case the date stamp is not giving us absolute precision, either. The hash of package contents could ensure that such version numbers are unique when when the code changes, but just a date stamp would be sufficient.

Alternatively, a `default` value could be provided to `use_scm_version`, so that `setuptools_scm` can still take care of the boilerplate exception handling when no VCS is available, and the user just has to provide whatever default suits their needs:

``` python
setuptools.setup(
    use_scm_version={
        'version_scheme': 'post-release',
        'default': '0+d'+datetime.date.today().strftime('%Y%m%d'),
    },
    setup_requires=['setuptools_scm'],
)
```

This is still very clean for the end user and gives them ultimate flexibility in what type of version number they are happy with for their purpose when VCS is not available.

**Comment by RonnyPfannschmidt** [comment]
We do crash and burn precisely because people did end up with wrong metadata on install from source

The example with 0 as fallback does not work for any package that is not a leaf in the dependency tree
Tracking the actual version as a fallback as well kind of defeats the purpose

There are already 2 standard formats to carry the version along - wheel and xdist

I don't consider supporting distribution formats that where stripped of the metadata explicitly a worthwhile goal, but its supportable using a fallback entry point similar to how the git-archive plugin works (which is limited to git archives of tagged versions)
I strongly recommend making such behaviour a plugin and would help with supporting APIs in core

**Comment by mrmachine** [comment]
Fair points. I'm not sure I follow the recommendation to create a plugin. I'm probably just not familiar enough with `setuptools` and `setuptools_scm`. Would that be a `setuptools_scm` plugin?

Could it work with a source directory, e.g. a working copy, copied into the docker context without the `.git` directory, or would it need to be an archive of some sort?

Could it be made to add feature described above, where I can provide any version number I like to use as a default when `setuptools_scm` fails, instead of crashing? The problem I want to avoid is having to try and predict the circumstances where `setuptools_scm` will crash, instead of being able to just run `setuptools_scm` in a `try/except` block because it is only actually downloaded and installed when `setup()` executes.

Like I said, right now I do this check by assuming it will crash if `.git` directory does not exist, but I don't like having to repeat this boilerplate code in every project we want to install into docker containers, and I imagine there could easily be other scenarios where `setuptools_scm` might crash and my package will fail to install.

If I could move that `try/except` into a plugin for `setuptools_scm`, that would be great. For example:

``` python
setuptools.setup(
    use_scm_version={'version_scheme': 'post-release'},
    setup_requires=[
        'setuptools_scm',
        'my_setuptools_scm_plugin',
    ],
    my_scm_plugin_fallback='0+d'+datetime.date.today().strftime('%Y%m%d'),
)
```

**Comment by RonnyPfannschmidt** [comment]
all you need is a entrypoint for `setuptools_scm.parse_scm_fallback`
pick a filename, use that as the entrypoint name and write a version to it and ensure that function returns that version + a dirty mark so the default date scheme will add a date

then you can either just put a `0` into it, or update it however you want - perhaps by adding a setuptools command, or jsust by a script, piping a version into it

however i would still suggest to just use a wheel instead, and have the wheel/build folders pip will use as  mounted folder while constructing the container so the intermediate files wont clutter the overlays

**Comment by RonnyPfannschmidt** [comment]
closing this one due to lack of further discussion, ping me if you think that decission was wrong

**Comment by ddelange** [comment]
For future reference (building only works in recent docker versions with buildkit support) you can avoid having `.git` in one of your layers by temporarily bind-mounting it during build time [[docs]](https://github.com/moby/buildkit/blob/v0.8.3/frontend/dockerfile/docs/syntax.md):

```dockerfile
RUN --mount=source=.git,target=.git,type=bind \
    pip install --no-cache-dir -e .
```

**Comment by maxnoe** [comment]
> For future reference (building only works in recent docker versions with buildkit support) you can avoid having .git in one of your layers by temporarily bind-mounting it during build time [[docs]](https://github.com/moby/buildkit/blob/v0.8.3/frontend/dockerfile/docs/syntax.md):

This will most likely have the side-effect of marking your setuptools_scm version as dirty, as you copy some needed files from the repo by hand into the Docker context, and then mount the `.git`. Git will see many files as deleted, thus the setuptools_scm version will be dirty.

**Comment by ddelange** [comment]
you can do a bind-mount of `.` (which includes `.git`) and then a pip install without `-e`. so you don't do any COPY, you won't be dirty, and only the necessary files will end up in the layer (in site-packages) 👍

**Comment by RonnyPfannschmidt** [comment]
Did you verify,?

**Comment by ddelange** [comment]
yes, here's a MRE.

you'll need a `rw` bind mount ([Allow writes on the mount. Written data will be discarded.](https://docs.docker.com/reference/dockerfile/#run---mounttypebind)) because pip will want to touch the `<pgk>.egg-info` directory:

```sh
git clone https://github.com/ddelange/mapply/
cd mapply
printf '
FROM python

WORKDIR /app

RUN --mount=type=bind,rw,source=.,target=/app \
    --mount=type=cache,target=/root/.cache/pip \
    pip install .

RUN python -c "import mapply"
' > ./Dockerfile
docker build .
```

**Comment by RonnyPfannschmidt** [comment]
Nice 

Well need a example that includes pip or uv cache mounts i believe

**Comment by ddelange** [comment]
the RUN command supports multiple `--mount` directives so you can add an additional cache mount if you like: `--mount=type=cache,target=/root/.cache/pip` or `--mount=type=cache,target=/root/.cache/uv` respectively. added to the MRE 👍
